src/query/storages/fuse/fuse/src/operations/mutation/compact_mutator/segment_compact_mutator.rs

```diff
@@ -237,10 +163,133 @@ impl TableMutator for SegmentCompactMutator {
             .commit_mutation(
                 &self.ctx,
                 self.compact_params.base_snapshot,
-                self.merged_segments_locations,
-                self.merged_segment_statistics,
+                self.compaction.segments_locations,
+                self.compaction.statistics,
                 abort_action,
             )
             .await
     }
 }
+
+// Segments compactor that preserver the order of ingestion.
+//
+// Since the order of segments( and the order of blocks as well) should be preserved,
+// if only segments of size "threshold" are allowed to be generated during compaction,
+// there might be cases that to compact one fragmented segment, a large amount of
+// non-fragmented segments have to be split into pieces and re-compacted.
+//
+// To avoid this "ripple effects", consecutive segments are allowed to be compacted into
+// a new segment, if the size of compacted segment is lesser than 2 * threshold (exclusive).
+
+pub struct SegmentCompactor<'a> {
+    // Size of compacted segment should be in range R == [threshold, 2 * threshold)
+    // within R, smaller one is preferred
+    threshold: u64,
+    // fragmented segment collected so far, it will be reset to empty if compaction occurs
+    fragmented_segments: Vec<(&'a SegmentInfo, Location)>,
+    // state which keep the number of blocks of all the fragmented segment collected so far,
+    // it will be reset to 0 if compaction occurs
+    accumulated_num_blocks: u64,
+    segment_writer: SegmentWriter<'a>,
+    // accumulated compaction state
+    compacted_state: SegmentCompactionState,
+}
+
+impl<'a> SegmentCompactor<'a> {
+    pub fn new(threshold: u64, segment_writer: SegmentWriter<'a>) -> Self {
+        Self {
+            threshold,
+            accumulated_num_blocks: 0,
+            fragmented_segments: vec![],
+            segment_writer,
+            compacted_state: Default::default(),
+        }
+    }
+
+    // accumulate one segment
+    pub async fn add(&mut self, segment_info: &'a SegmentInfo, location: Location) -> Result<()> {
```

Comment:
So we cannot ensure the exact number of blocks in a segment is   threshold.

Reply:
u r right! the reason behind this:

>to avoid the ripple effects which may lead to a large number of consecutive compacted segments being split and merged again and again, for even a single fragmented segment, during compaction, segments of size [threshold, 2 * threshold) are allowed to be generated.